### ISSUE TITLE
Backup deletion on SSH server broken ?

### ISSUE THREAD

**Issue by beat** [virtualmin/virtualmin-gpl#312]
# Description:

Backup is set to delete after 7 days, and this is confirmed in the output of scheduled backup log:

```
Deleting backups from /mnt/backups/m2/m2-%Y-%m-%d on SSH server backups.mydomain.com
older than 7 days ..
.. no backups to delete were found
```

but there are backups older than 7 days (SSH through SSH key file and not password anymore)

```
root@backups:/mnt/backups/m2# ls -al
total 52
drwxr-xr-x 13 m2.backups backups 4096 Oct  5 11:17 .
drwxr-x--- 21 backups      backups 4096 Sep 30 22:26 ..
drwxr-xr-x  2 m2.backups backups 4096 Sep 24 04:02 m2-2021-09-24
drwxr-xr-x  2 m2.backups backups 4096 Sep 25 04:02 m2-2021-09-25
drwxr-xr-x  2 m2.backups backups 4096 Sep 26 04:03 m2-2021-09-26
drwxr-xr-x  2 m2.backups backups 4096 Sep 28 04:56 m2-2021-09-28
drwxr-xr-x  2 m2.backups backups 4096 Sep 29 04:59 m2-2021-09-29
drwxr-xr-x  2 m2.backups backups 4096 Sep 30 04:59 m2-2021-09-30
drwxr-xr-x  2 m2.backups backups 4096 Oct  1 04:26 m2-2021-10-01
drwxr-xr-x  2 m2.backups backups 4096 Oct  2 04:27 m2-2021-10-02
drwxr-xr-x  2 m2.backups backups 4096 Oct  3 17:26 m2-2021-10-03
drwxr-xr-x  2 m2.backups backups 4096 Oct  4 04:56 m2-2021-10-04
drwxr-xr-x  2 m2.backups backups 4096 Oct  5 04:58 m2-2021-10-05
```

I see the same behavior on all my virtualmin servers, independantly of number of backups to hold. That fills up the backup server up to disk full.

**Comment by beat** [comment]
@jcameron those directories contain just the normal files:

- for each domain: domain.tar.gz, domain.tar.gz.dom, domain.tar.gz.info
- and same for the virtualmin configs backup: virtualmin.tar.gz, virtualmin.tar.gz.dom, virtualmin.tar.gz.info

Dates of directories and of all files in them correspond to the backup date.

**Comment by jcameron** [comment]
Does this user you're using for the backups have the `scponly` shell configured? This only allows uploading and downloading of files, and so isn't compatible with purging old backups.

**Comment by beat** [comment]
Yes, they have the scponly shell configured (as was already the case on the backup system in Ubuntu 16.04).

But deletes were working then (from Ubuntu 16.04 clients and from Ubuntu 20.04 client): E.g.:

```
Uploading archive to SSH server backup.example.com ..
.. done

16 servers backed up successfully, 0 had errors.
8 Virtualmin configuration settings backed up successfully.
Deleting backups from /mnt/backups/m2/m2-%Y-%m-%d on SSH server backup.example.com older than 32 days ..
    Deleting file /mnt/backups/m2/m2-2021-05-01 via SSH, which is 32 days old ..
    .. deleted 4 KiB.

.. deleted 1 old backups, and skipped 35 that were not old enough

Running post-backup command ..
.. done

```

The "4 KiB" size was definitely wrong, as that was the "size" of the directory and not total size of all files inside the directory, but it worked as long as the backup server was Ubuntu 16.04 (all users receiving the backup always had "Email and scp" as "other user permissions" (shell /usr/bin/scponly).

As it was working fine, I don't wish to give full user SSH access to the backup server to all servers, so kept scponly.

What's wierd is that it **was** working with ubuntu 16.04 and 20.04 clients and SSH with passwords instead of keys and 16.04 backup server, but not anymore with the backup server upgraded to 20.04 and using a SSH key file per client instead of a SSH password per client.

Also wierd that the script now just says "no backups to delete were found
", no error messages that it couldn't e.g. list or delete the old files ?

**Comment by jcameron** [comment]
That's a good suggestion - let me see if we can display a better error message in this case, and perhaps use a different subject line if the backup succeeded but deletion failed.

**Comment by beat** [comment]
Awesome! Thank you Jamie!

Btw, trying to secure the individual ssh (non-scponly) accesses to the backup server, but didn't find any functionality in Virtualmin to chroot-jail individual users of a virtual server, nor to chroot-jail existing virtual servers' main users (but saw the setting for new virtual servers, so only solution through Virtualmin seems to be to set the "Chroot jail new domain Unix users" in server template/Administration User, then to create new virtual server one per server to backup, with domain unix user with then chrooted ssh access, and find a way to link to the backup disk individual directory). Any hint?

**Comment by jcameron** [comment]
You can do this from the command line, with a command like `virtualmin modify-domain --domain example.com --enable-jail`

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
